blog/2022/05-10-resoto-graph-traversal/index.md

```diff
@@ -0,0 +1,119 @@
+---
+authors: [matthias]
+tags: [graph, search, digitalocean]
+---
+
+# A walk in the graph
+
+Resoto uses a [directed graph](https://en.wikipedia.org/wiki/Directed_graph) to represent your infrastructure resources as nodes and relationships between them as edges. It even stores the data in a graph database, being able to query the graph efficiently. Why should I care, you ask? To answer this question, I would like to present an example that demonstrates the power of this approach.
+
+The following diagram shows an excerpt of the graph model for DigitalOcean droplets; the compute resource in DigitalOcean (similar to EC2 instance in AWS or Instance in GCP). The full model is available in our [reference section](/docs/reference/data-models/digitalocean). Please note that this example uses DigitalOcean, but all examples could be adapted and used also for [AWS](/docs/reference/data-models/aws) or [GCP](/docs/reference/data-models/gcp).
+
+![DigitalOcean Droplet](img/digitalocean_droplet.svg)
+
+This diagram shows how the different kinds of resources are connected to each other in the graph. It does not mean that all of these resources have to be available - it only shows the static structure of a possible graph. A droplet, for example, can have volumes, snapshots, or floating IPs, which are direct successors of the node. If we want to know which volumes are attached to a droplet, we will walk the graph outbound precisely by one step. Since this could yield volumes, snapshots, and floating IPs, as seen in the diagram, we need to filter the resulting list to only keep the volumes. Let's assume we have a specific droplet with id `289061882`, where we want to know all attached volumes:
+
+```bash
+> search is(digitalocean_droplet) and id=289061882 --> is(volume)
+// highlight-start
+​kind=digitalocean_volume, name=pv2, age=2mo7d, account=10225075, region=Amsterdam 3
+​kind=digitalocean_volume, name=pv1, age=2mo7d, account=10225075, region=Amsterdam 3
+// highlight-end
+```
+
+This search finds the DigitalOcean droplet with the related id and then walks the graph outbound by one step, as defined by `-->` , and filters the resulting volume list. Other resources are pointing towards our droplet. By looking at the graph structure, we see it could be load balancers, kubernetes clusters, firewalls, or images. To get all the resources that point to this resource, we need to walk the graph inbound - which means in the opposite direction of the arrow in the graph structure diagram. Resoto uses the reversed arrow `<--` to walk the graph inbound. We use the same filter as before - this time, we walk the graph inbound and count the number of resources by kind.
```

Comment:
```suggestion
This search finds the DigitalOcean droplet with the related id, then walks the graph outbound by one step, as defined by `-->` , and filters the resulting volume list. Other resources are pointing towards our droplet. By looking at the graph structure, we see it could be load balancers, kubernetes clusters, firewalls, or images. To get all the resources that point to this resource, we need to walk the graph inbound - which means in the opposite direction of the arrow in the graph structure diagram. Resoto uses the left facing arrow `<--` to walk the graph inbound. We use the same filter as before but this time, we walk the graph inbound and count the number of resources by kind.
```